lib/IRGen/GenBuiltin.cpp

```diff
@@ -1537,3 +1539,682 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+              bool isOutlined, SILType T) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.assign(IGF, explosion, elementAddr, isOutlined,
+                                     eltTy);
+                    });
+  }
+
+  void initialize(IRGenFunction &IGF, Explosion &explosion, Address addr,
+                  bool isOutlined) const override {
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.initialize(IGF, explosion, elementAddr, isOutlined);
+                    });
+  }
+  
+  void reexplode(Explosion &sourceExplosion,
+                 Explosion &targetExplosion) const override {
+    eachElement([&]{
+      Element.reexplode(sourceExplosion, targetExplosion);
+    });
+  }
+  
+  void copy(IRGenFunction &IGF,
+            Explosion &sourceExplosion,
+            Explosion &targetExplosion,
+            Atomicity atomicity) const override {
+    eachElement([&]{
+      Element.copy(IGF, sourceExplosion, targetExplosion, atomicity);
+    });
+  }
+  
+  void consume(IRGenFunction &IGF, Explosion &explosion,
+               Atomicity atomicity,
+               SILType T) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElement([&]{
+      Element.consume(IGF, explosion, atomicity, eltTy);
+    });
+  }
+
+  void fixLifetime(IRGenFunction &IGF,
+                   Explosion &explosion) const override {
+    eachElement([&]{
+      Element.fixLifetime(IGF, explosion);
+    });
+  }
+
+  template<typename Body>
+  void eachElementOffset(Body &&body) const {
+    for (unsigned i = 0; i < ArraySize; ++i) {
+      body(i * Element.getFixedStride().getValue());
+    }
+  }
+  
+  void packIntoEnumPayload(IRGenModule &IGM,
+                           IRBuilder &builder,
+                           EnumPayload &payload,
+                           Explosion &sourceExplosion,
+                           unsigned offset) const override {
+    eachElementOffset([&](unsigned eltByteOffset){
+      Element.packIntoEnumPayload(IGM, builder, payload,
+                                  sourceExplosion,
+                                  offset + eltByteOffset * 8);
+    });
+  }
+  
+  void unpackFromEnumPayload(IRGenFunction &IGF,
+                             const EnumPayload &payload,
+                             Explosion &targetExplosion,
+                             unsigned offset) const override {
+    eachElementOffset([&](unsigned eltByteOffset){
+      Element.unpackFromEnumPayload(IGF, payload,
+                                    targetExplosion,
+                                    offset + eltByteOffset * 8);
+    });
+  }
+  
+  void addToAggLowering(IRGenModule &IGM, SwiftAggLowering &lowering,
+                        Size offset) const override {
+    eachElementOffset([&](unsigned eltByteOffset){
```

Comment:
At what point do you check that `offset * elementSize` doesn't overflow?  And actually, the same question applies to the dynamic cases, right?

Reply:
Hmm, we don't anywhere else we do type layout, but you're right we probably should here since it's a lot easier to induce a size overflow now.

Reply:
It probably doesn't make any sense to treat huge arrays as loadable at any level, so I put a cap at 4096 after which we lower the type as address-only in SIL, and Fixed but not Loadable in IRGen.

For dynamic cases, I'm wondering if it'd be sufficient to have the standard library type do a precondition check that the value wouldn't be larger than address space before attempting initialization. If you can never construct such a value to begin with, then the value witness operations should never occur. Though I suppose it would be possible to have a type like

```
struct Foo<let N: Int> {
  var buffer: Vector<N, Int>
  var somethingElse: Int

  init() {
    self.somethingElse = 42
    self.buffer = initialBuffer()
  }
}
```

and if an attacker got control of N, they could trick the `init()` call into writing 42 somewhere bad before the `self.buffer` initializer got a chance to trap. Maybe we could cap the value witness table size of these array types at some practical limit (maybe 2^31-1 for 32-bit platforms, and 2^55-1 on 64-bit platforms), so that if a huge array shows up as part of an aggregate, the offsets for following fields have some cushion before they wrap around?
```

Reply:
I guess we should probably step back and get this right in general.  Semantically, we should really be checking legality in the initializer of every type, and it's just that fixed-size types can emit that check statically, hopefully by removing it.  Fixed-size arrays obviously make this much easier to violate, but it's always been the right rule.  And fixed-size arrays make it easier to violate in all cases, not just when the size of the fixed-size array type itself overflows.  So I guess there should be a flag in the VWT for illegal types, and struct/enum/class dynamic layout needs to detect this and set the flag appropriately.

If the semantics are that illegal types are well-formed but invalid to instantiate, ideally they should have zero size. That might screw up extra-inhabitant enum layout, though, since it would make the entire case go away.  Do we ever use extra-inhabitant layout strategies based on non-fixed enum cases, e.g. by knowing that the payload is a tuple whose first element has a fixed layout?

Reply:
> If the semantics are that illegal types are well-formed but invalid to instantiate, ideally they should have zero size. That might screw up extra-inhabitant enum layout, though, since it would make the entire case go away. Do we ever use extra-inhabitant layout strategies based on non-fixed enum cases, e.g. by knowing that the payload is a tuple whose first element has a fixed layout?

Yeah, for tuples and structs, we pick the element with the most extra inhabitants to provide extra inhabitants for the aggregate.

If we can properly track type legality, and make such types treated as zero-size, that would have the nice theoretical benefit that an illegal type could be substituted into an enum payload while still allowing the valid payloads of the enum to be practically usable. I wonder if we could retroactively track `Never` as being illegal too; that might "break" the size we expect of structs with unconditional Never fields, but those structs couldn't actually be instantiated…

Reply:
Yeah, I think we can make this work, we just have to find a rule that's consistent with our current assumptions about type layout.  In particular, consider a type like this:

```
  struct A<T> {
    var x: SomeClassReference
    var y: T
  }
```

When we're laying out `Optional<A<T>>` generically, do we always assume `A<T>` has extra inhabitants because we can see that `x` does, such that e.g. constructing `.none` can just write a null pointer (as the first XI) into that field?  If so, then the illegal-type layout rule probably has to stay consistent with that — whatever fields are assumed to always exist have to continue to exist, and it's only the other fields that can be removed when we recognize an illegal type.

Ultimately, I think we're deciding between three rules here:
1. Product types with any illegal fields are zero-sized overall.
2. All fields *following* illegal fields in product types are zero-sized. (Probably legal only if we don't make XI assumptions
3. Product types just treat illegal fields as zero-sized but don't otherwise change layout.

(1) only works if we don't make any XI assumptions about known fields when laying out enums of generic non-fixed product types. (2) only works if those XI assumptions are only made about in the fixed-size prefix of the product. (3) should work in any case, but we might be in deep trouble if illegality starts in the product itself (i.e. all of the fields are legal, but the product layout overflows).

Reply:
> When we're laying out Optional<A<T>> generically, do we always assume A<T> has extra inhabitants because we can see that x does, such that e.g. constructing .none can just write a null pointer (as the first XI) into that field? If so, then the illegal-type layout rule probably has to stay consistent with that — whatever fields are assumed to always exist have to continue to exist, and it's only the other fields that can be removed when we recognize an illegal type.

It looks like dynamic type layout for `A` dynamically picks the field with the most extra inhabitants:

```
  // We have extra inhabitants if any element does. Use the field with the most.
  unsigned extraInhabitantCount = 0;
  for (unsigned i = 0; i < numFields; ++i) {
    unsigned fieldExtraInhabitantCount =
        fieldTypes[i]->getNumExtraInhabitants();
    if (fieldExtraInhabitantCount > extraInhabitantCount) {
      extraInhabitantCount = fieldExtraInhabitantCount;
    }
  }
```

So we should be able to make it so that a product with an illegal field itself becomes fully illegal.